/*!
 * @file
 *
 * @section LICENSE
 *
 * Copyright (C) 2017 by the Georgia Tech Research Institute (GTRI)
 *
 * This file is part of SCRIMMAGE.
 *
 *   SCRIMMAGE is free software: you can redistribute it and/or modify it under
 *   the terms of the GNU Lesser General Public License as published by the
 *   Free Software Foundation, either version 3 of the License, or (at your
 *   option) any later version.
 *
 *   SCRIMMAGE is distributed in the hope that it will be useful, but WITHOUT
 *   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 *   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 *   License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with SCRIMMAGE.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Kevin DeMarco <kevin.demarco@gtri.gatech.edu>
 * @author Eric Squires <eric.squires@gtri.gatech.edu>
 * @date 31 July 2017
 * @version 0.1.0
 * @brief Brief file description.
 * @section DESCRIPTION
 * A Long description goes here.
 *
 */
syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.syllo.scrimmage";

import "scrimmage/proto/State.proto";

package scrimmage_proto;

service ExternalControl {
  rpc SendEnvironment (Environment) returns (Action) {}
  rpc SendActionResult (ActionResult) returns (Action) {}
}

message DiscreteSpace {
    int32 minimum = 1;
    int32 maximum = 2;
}

message ContinuousSpace {
    double minimum = 1;
    double maximum = 2;
}

message Space {
    repeated DiscreteSpace discrete = 1;
    repeated ContinuousSpace continuous = 2;
}

message Environment {
    Space action_space = 1;
    Space observation_space = 2;
    double min_reward = 3;
    double max_reward = 4;
}

message Action {
    repeated State states = 1;
}

message ActionResult {
    SpaceSample observation = 1;
    double reward = 2;
    bool done = 3;
}
